File: kaji/sdk/src/integrations/fixed_origin.py
"""Bounded HTTP for provider-owned, constructor-fixed HTTPS origins."""

from __future__ import annotations

import asyncio
from collections.abc import Callable, Mapping
from contextlib import suppress
from dataclasses import dataclass
import math
import time
from types import MappingProxyType
from typing import Literal
from urllib.parse import unquote, urlsplit

import httpx

from kaji.integrations.errors import IntegrationPolicyError, IntegrationTransportError
from kaji.infra.observability.protocols import (
    MetricsSink,
    NOOP_METRICS,
    NOOP_TRACE,
    TraceSink,
    record_metric,
    start_span,
)
from kaji.runtime.agents.cancellation import CancelledError
from kaji.runtime.context import ToolExecutionContext


_DEFAULT_MAX_RESPONSE_BYTES = 1_048_576
_MAX_RESPONSE_HEADER_FIELDS = 64
_MAX_RESPONSE_HEADER_BYTES = 64 * 1024
_FORBIDDEN_REQUEST_HEADERS = frozenset(
    {
        "connection",
        "content-length",
        "host",
        "keep-alive",
        "proxy-authenticate",
        "proxy-authorization",
        "proxy-connection",
        "te",
        "trailer",
        "transfer-encoding",
        "upgrade",
    }
)


@dataclass(frozen=True, slots=True)
class _FixedOriginPolicy:
    origin: str
    integration: Literal["github", "gmail"]
    timeout_seconds: float = 10.0
    max_response_bytes: int = _DEFAULT_MAX_RESPONSE_BYTES
    allowed_methods: tuple[str, ...] = ("GET", "POST")


@dataclass(frozen=True, slots=True)
class IntegrationResponse:
    status: int
    headers: Mapping[str, str]
    body: bytes


def _policy_error() -> IntegrationPolicyError:
    return IntegrationPolicyError()


def _response_limit() -> IntegrationTransportError:
    return IntegrationTransportError(
        "INTEGRATION_RESPONSE_LIMIT", "response_limit_exceeded"
    )


def _validated_origin(value: str) -> str:
    if not isinstance(value, str):
        raise _policy_error()
    parsed = urlsplit(value)
    if (
        parsed.scheme != "https"
        or not parsed.hostname
        or parsed.username is not None
        or parsed.password is not None
        or parsed.path not in {"", "/"}
        or parsed.query
        or parsed.fragment
    ):
        raise _policy_error()
    origin = f"https://{parsed.hostname.lower()}"
    if parsed.port is not None:
        origin += f":{parsed.port}"
    return origin


def _validate_policy(policy: _FixedOriginPolicy) -> tuple[str, frozenset[str]]:
    origin = _validated_origin(policy.origin)
    if (
        isinstance(policy.timeout_seconds, bool)
        or not isinstance(policy.timeout_seconds, (int, float))
        or not math.isfinite(float(policy.timeout_seconds))
        or policy.timeout_seconds <= 0
    ):
        raise _policy_error()
    if (
        isinstance(policy.max_response_bytes, bool)
        or not isinstance(policy.max_response_bytes, int)
        or policy.max_response_bytes < 1
    ):
        raise _policy_error()
    methods = tuple(policy.allowed_methods)
    if not methods or any(method not in {"GET", "POST"} for method in methods):
        raise _policy_error()
    return origin, frozenset(methods)


def _validated_target(origin: str, path_and_query: str) -> str:
    if (
        not isinstance(path_and_query, str)
        or not path_and_query.startswith("/")
        or path_and_query.startswith("//")
        or "\\" in path_and_query
        or "#" in path_and_query
    ):
        raise _policy_error()
    path = path_and_query.split("?", 1)[0]
    try:
        decoded_path = unquote(path, errors="strict")
    except (UnicodeDecodeError, ValueError):
        raise _policy_error() from None
    if decoded_path.startswith("//") or "\\" in decoded_path:
        raise _policy_error()
    target = origin + path_and_query
    parsed = urlsplit(target)
    if f"{parsed.scheme}://{parsed.netloc}" != origin:
        raise _policy_error()
    return target


def _validated_request_headers(headers: Mapping[str, str]) -> dict[str, str]:
    if not isinstance(headers, Mapping):
        raise _policy_error()
    result: dict[str, str] = {}
    for name, value in headers.items():
        if not isinstance(name, str) or not isinstance(value, str):
            raise _policy_error()
        normalized = name.strip().lower()
        if (
            not normalized
            or normalized in _FORBIDDEN_REQUEST_HEADERS
            or normalized.startswith("proxy-")
            or "\r" in name
            or "\n" in name
            or "\r" in value
            or "\n" in value
        ):
            raise _policy_error()
        result[normalized] = value
    return result


def _bounded_response_headers(headers: httpx.Headers) -> Mapping[str, str]:
    items = headers.multi_items()
    if len(items) > _MAX_RESPONSE_HEADER_FIELDS:
        raise _response_limit()
    size = sum(len(name.encode()) + len(value.encode()) for name, value in items)
    if size > _MAX_RESPONSE_HEADER_BYTES:
        raise _response_limit()
    content_lengths = headers.get_list("content-length")
    if content_lengths:
        if len(content_lengths) != 1 or not content_lengths[0].isdigit():
            raise _response_limit()
    return MappingProxyType({name.lower(): value for name, value in items})


class FixedOriginClient:
    """Pooled client whose production construction fixes one provider origin."""

    def __init__(
        self,
        policy: _FixedOriginPolicy,
        *,
        transport: httpx.AsyncBaseTransport | None = None,
        metrics_sink: MetricsSink = NOOP_METRICS,
        trace_sink: TraceSink = NOOP_TRACE,
        monotonic_now: Callable[[], float] = time.monotonic,
    ) -> None:
        self._origin, self._allowed_methods = _validate_policy(policy)
        self._policy = policy
        self._metrics_sink = metrics_sink
        self._trace_sink = trace_sink
        self._monotonic_now = monotonic_now
        self._client = httpx.AsyncClient(
            transport=transport,
            timeout=None,
            follow_redirects=False,
            trust_env=False,
        )

    @classmethod
    def for_github(
        cls,
        *,
        metrics_sink: MetricsSink = NOOP_METRICS,
        trace_sink: TraceSink = NOOP_TRACE,
    ) -> FixedOriginClient:
        return cls(
            _FixedOriginPolicy("https://api.github.com", "github"),
            metrics_sink=metrics_sink,
            trace_sink=trace_sink,
        )

    @classmethod
    def for_gmail(
        cls,
        *,
        metrics_sink: MetricsSink = NOOP_METRICS,
        trace_sink: TraceSink = NOOP_TRACE,
    ) -> FixedOriginClient:
        return cls(
            _FixedOriginPolicy("https://gmail.googleapis.com", "gmail"),
            metrics_sink=metrics_sink,
            trace_sink=trace_sink,
        )

    @classmethod
    def _for_test(
        cls,
        origin: str,
        *,
        transport: httpx.AsyncBaseTransport,
        timeout_seconds: float = 10.0,
        max_response_bytes: int = _DEFAULT_MAX_RESPONSE_BYTES,
        allowed_methods: tuple[str, ...] = ("GET", "POST"),
        integration: Literal["github", "gmail"] = "github",
        metrics_sink: MetricsSink = NOOP_METRICS,
        trace_sink: TraceSink = NOOP_TRACE,
        monotonic_now: Callable[[], float] = time.monotonic,
    ) -> FixedOriginClient:
        return cls(
            _FixedOriginPolicy(
                origin,
                integration,
                timeout_seconds=timeout_seconds,
                max_response_bytes=max_response_bytes,
                allowed_methods=allowed_methods,
            ),
            transport=transport,
            metrics_sink=metrics_sink,
            trace_sink=trace_sink,
            monotonic_now=monotonic_now,
        )

    async def request(
        self,
        path_and_query: str,
        *,
        method: str,
        headers: Mapping[str, str],
        body: bytes | None,
        context: ToolExecutionContext,
    ) -> IntegrationResponse:
        started = self._monotonic_now()
        operation_label = "read" if method == "GET" else "mutation"
        span = start_span(
            self._trace_sink,
            "kaji.integration.request",
            {
                "integration.name": self._policy.integration,
                "integration.operation": operation_label,
                "http.status_family": "none",
            },
        )
        outcome = "error"
        try:
            if method not in self._allowed_methods or (
                body is not None and not isinstance(body, bytes)
            ):
                raise _policy_error()
            target = _validated_target(self._origin, path_and_query)
            request_headers = _validated_request_headers(headers)
            context.cancellation_token.raise_if_cancelled()
            remaining = self._policy.timeout_seconds
            if context.deadline_monotonic is not None:
                remaining = min(
                    remaining,
                    context.deadline_monotonic - time.monotonic(),
                )
            if remaining <= 0:
                raise TimeoutError("Integration request timed out")

            request_task = asyncio.create_task(
                self._request_once(target, method, request_headers, body)
            )
            cancelled = asyncio.create_task(context.cancellation_token.wait())
            try:
                done, _ = await asyncio.wait(
                    {request_task, cancelled},
                    timeout=remaining,
                    return_when=asyncio.FIRST_COMPLETED,
                )
                if request_task in done:
                    response = request_task.result()
                    span.set_attribute(
                        "http.status_family",
                        (
                            f"{response.status // 100}xx"
                            if 100 <= response.status <= 599
                            else "none"
                        ),
                    )
                    outcome = "success" if 200 <= response.status < 300 else "error"
                    return response
                request_task.cancel()
                with suppress(asyncio.CancelledError):
                    await request_task
                if cancelled in done or context.cancellation_token.is_cancelled:
                    raise CancelledError("Integration request cancelled")
                raise TimeoutError("Integration request timed out")
            finally:
                cancelled.cancel()
                with suppress(asyncio.CancelledError):
                    await cancelled
        except BaseException as error:
            if isinstance(error, (asyncio.CancelledError, CancelledError)):
                outcome = "cancelled"
            span.record_error(error)
            raise
        finally:
            record_metric(
                self._metrics_sink,
                "kaji.integration.request_ms",
                max(0.0, (self._monotonic_now() - started) * 1_000),
                integration=self._policy.integration,
                operation=operation_label,
                outcome=outcome,
            )
            span.end()

    async def _request_once(
        self,
        target: str,
        method: str,
        headers: Mapping[str, str],
        body: bytes | None,
    ) -> IntegrationResponse:
        async with self._client.stream(
            method,
            target,
            headers=headers,
            content=body,
        ) as response:
            bounded_headers = _bounded_response_headers(response.headers)
            if 300 <= response.status_code < 400:
                raise IntegrationTransportError(
                    "INTEGRATION_REDIRECT_REJECTED", "redirect_rejected"
                )
            content_length = bounded_headers.get("content-length")
            if (
                content_length is not None
                and int(content_length) > self._policy.max_response_bytes
            ):
                raise _response_limit()
            chunks: list[bytes] = []
            size = 0
            async for chunk in response.aiter_bytes():
                size += len(chunk)
                if size > self._policy.max_response_bytes:
                    raise _response_limit()
                chunks.append(chunk)
            return IntegrationResponse(
                status=response.status_code,
                headers=bounded_headers,
                body=b"".join(chunks),
            )

    async def aclose(self) -> None:
        await self._client.aclose()

File: kaji/ts/src/integrations/fixed-origin.ts
import { Agent, request as httpsRequest } from "node:https";

type HeaderInput = ConstructorParameters<typeof Headers>[0];

import type { ToolExecutionContext } from "@/runtime/context";
import type { BoundedResponse } from "@/integrations/safe-fetch";
import { IntegrationPolicyError, IntegrationTransportError } from "@/integrations/errors";
import {
  NOOP_METRICS,
  NOOP_TRACE,
  recordMetric,
  startSpan,
  type MetricsSink,
  type TraceSink,
} from "@/observability";

const DEFAULT_TIMEOUT_MS = 10_000;
const DEFAULT_MAX_RESPONSE_BYTES = 1_048_576;
const MAX_RESPONSE_HEADER_FIELDS = 64;
const MAX_RESPONSE_HEADER_BYTES = 64 * 1024;
const FORBIDDEN_REQUEST_HEADERS = new Set([
  "connection",
  "content-length",
  "host",
  "keep-alive",
  "proxy-authenticate",
  "proxy-authorization",
  "proxy-connection",
  "te",
  "trailer",
  "transfer-encoding",
  "upgrade",
]);

interface FixedOriginPolicy {
  readonly origin: URL;
  readonly integration: "github" | "gmail";
  readonly timeoutMs?: number;
  readonly maxResponseBytes?: number;
  readonly allowedMethods?: readonly ("GET" | "POST")[];
}

interface FixedOriginObservability {
  readonly metricsSink?: MetricsSink;
  readonly traceSink?: TraceSink;
  readonly monotonicNow?: () => number;
}

export interface FixedOriginRequester {
  request(
    pathAndQuery: string,
    init: Readonly<{
      method: "GET" | "POST";
      headers: HeaderInput;
      body?: Uint8Array;
    }>,
    context: ToolExecutionContext,
  ): Promise<BoundedResponse>;
}

interface FixedOriginTransportInit {
  readonly method: "GET" | "POST";
  readonly headers: Readonly<Record<string, string>>;
  readonly body?: Uint8Array;
  readonly signal: AbortSignal;
}

/** @internal Relative source tests only; absent from the package subpath. */
export interface FixedOriginTestResponse {
  readonly status: number;
  readonly headers: readonly (readonly [string, string])[];
  readonly body: AsyncIterable<Uint8Array>;
  readonly close: () => void;
}

/** @internal Relative source tests only; absent from the package subpath. */
export interface FixedOriginTestTransport {
  request(url: URL, init: FixedOriginTransportInit): Promise<FixedOriginTestResponse>;
  close?(): void;
}

function policyError(): IntegrationPolicyError {
  return new IntegrationPolicyError();
}

function responseLimit(): IntegrationTransportError {
  return new IntegrationTransportError("INTEGRATION_RESPONSE_LIMIT", "response_limit_exceeded");
}

function positiveInteger(value: number | undefined, fallback: number): number {
  const resolved = value ?? fallback;
  if (!Number.isSafeInteger(resolved) || resolved < 1) throw policyError();
  return resolved;
}

function validatedOrigin(input: URL): URL {
  const origin = new URL(input.href);
  if (
    origin.protocol !== "https:" ||
    origin.username !== "" ||
    origin.password !== "" ||
    origin.pathname !== "/" ||
    origin.search !== "" ||
    origin.hash !== ""
  ) {
    throw policyError();
  }
  return origin;
}

function validatedPath(origin: URL, pathAndQuery: string): URL {
  if (
    typeof pathAndQuery !== "string" ||
    !pathAndQuery.startsWith("/") ||
    pathAndQuery.startsWith("//") ||
    pathAndQuery.includes("\\") ||
    pathAndQuery.includes("#")
  ) {
    throw policyError();
  }
  const rawPath = pathAndQuery.split("?", 1)[0]!;
  let decodedPath: string;
  try {
    decodedPath = decodeURIComponent(rawPath);
  } catch {
    throw policyError();
  }
  if (decodedPath.startsWith("//") || decodedPath.includes("\\")) throw policyError();
  const target = new URL(pathAndQuery, origin);
  if (target.origin !== origin.origin) throw policyError();
  return target;
}

function validatedRequestHeaders(input: HeaderInput): Readonly<Record<string, string>> {
  let headers: Headers;
  try {
    headers = new Headers(input);
  } catch {
    throw policyError();
  }
  const result: Record<string, string> = {};
  for (const [name, value] of headers) {
    const normalized = name.toLowerCase();
    if (FORBIDDEN_REQUEST_HEADERS.has(normalized) || normalized.startsWith("proxy-")) {
      throw policyError();
    }
    result[normalized] = value;
  }
  return Object.freeze(result);
}

function boundedResponseHeaders(
  input: readonly (readonly [string, string])[],
): Readonly<Record<string, string>> {
  if (input.length > MAX_RESPONSE_HEADER_FIELDS) throw responseLimit();
  const encoder = new TextEncoder();
  let size = 0;
  const result: Record<string, string> = {};
  const contentLengths: string[] = [];
  for (const [name, value] of input) {
    size += encoder.encode(name).byteLength + encoder.encode(value).byteLength;
    if (size > MAX_RESPONSE_HEADER_BYTES) throw responseLimit();
    const normalized = name.toLowerCase();
    if (normalized === "content-length") contentLengths.push(value);
    result[normalized] = value;
  }
  if (
    contentLengths.length > 1 ||
    (contentLengths.length === 1 && !/^\d+$/.test(contentLengths[0]!))
  ) {
    throw responseLimit();
  }
  return Object.freeze(result);
}

function safeAbort(name: "AbortError" | "TimeoutError"): DOMException {
  return new DOMException(
    name === "AbortError" ? "Integration request cancelled" : "Integration request timed out",
    name,
  );
}

function requestSignal(context: ToolExecutionContext, timeoutMs: number) {
  if (context.signal.aborted) throw safeAbort("AbortError");
  const remaining =
    context.deadlineMonotonicMs === undefined
      ? timeoutMs
      : Math.min(timeoutMs, context.deadlineMonotonicMs - performance.now());
  if (remaining <= 0) throw safeAbort("TimeoutError");
  const controller = new AbortController();
  const onAbort = () => controller.abort(safeAbort("AbortError"));
  context.signal.addEventListener("abort", onAbort, { once: true });
  const timer = setTimeout(() => controller.abort(safeAbort("TimeoutError")), remaining);
  return {
    signal: controller.signal,
    dispose() {
      clearTimeout(timer);
      context.signal.removeEventListener("abort", onAbort);
    },
  };
}

async function withAbort<T>(operation: Promise<T>, signal: AbortSignal): Promise<T> {
  if (signal.aborted) throw signal.reason;
  return new Promise<T>((resolve, reject) => {
    const onAbort = () => {
      cleanup();
      reject(signal.reason);
    };
    const cleanup = () => signal.removeEventListener("abort", onAbort);
    signal.addEventListener("abort", onAbort, { once: true });
    void operation.then(
      (value) => {
        cleanup();
        resolve(value);
      },
      (error: unknown) => {
        cleanup();
        reject(error);
      },
    );
  });
}

class NodeHttpsTransport implements FixedOriginTestTransport {
  private readonly agent = new Agent({ keepAlive: true });

  request(url: URL, init: FixedOriginTransportInit): Promise<FixedOriginTestResponse> {
    return new Promise((resolve, reject) => {
      const request = httpsRequest(
        url,
        { method: init.method, headers: init.headers, agent: this.agent },
        (response) => {
          const headers: Array<readonly [string, string]> = [];
          for (let index = 0; index < response.rawHeaders.length; index += 2) {
            headers.push([response.rawHeaders[index]!, response.rawHeaders[index + 1] ?? ""]);
          }
          resolve({
            status: response.statusCode ?? 0,
            headers,
            body: response,
            close: () => response.destroy(),
          });
        },
      );
      const onAbort = () => request.destroy(init.signal.reason);
      init.signal.addEventListener("abort", onAbort, { once: true });
      request.once("error", reject);
      request.once("close", () => init.signal.removeEventListener("abort", onAbort));
      request.end(init.body);
    });
  }

  close(): void {
    this.agent.destroy();
  }
}

class FixedOriginRequesterImpl implements FixedOriginRequester {
  private readonly origin: URL;
  private readonly timeoutMs: number;
  private readonly maxResponseBytes: number;
  private readonly allowedMethods: ReadonlySet<"GET" | "POST">;
  private readonly integration: "github" | "gmail";
  private readonly metrics: MetricsSink;
  private readonly trace: TraceSink;
  private readonly monotonicNow: () => number;
  private closed = false;

  constructor(
    policy: FixedOriginPolicy,
    private readonly transport: FixedOriginTestTransport,
    observability: FixedOriginObservability = {},
  ) {
    this.origin = validatedOrigin(policy.origin);
    this.integration = policy.integration;
    this.timeoutMs = positiveInteger(policy.timeoutMs, DEFAULT_TIMEOUT_MS);
    this.maxResponseBytes = positiveInteger(policy.maxResponseBytes, DEFAULT_MAX_RESPONSE_BYTES);
    const allowedMethods = policy.allowedMethods ?? ["GET", "POST"];
    if (
      allowedMethods.length === 0 ||
      allowedMethods.some((method) => method !== "GET" && method !== "POST")
    ) {
      throw policyError();
    }
    this.allowedMethods = new Set(allowedMethods);
    this.metrics = observability.metricsSink ?? NOOP_METRICS;
    this.trace = observability.traceSink ?? NOOP_TRACE;
    this.monotonicNow = observability.monotonicNow ?? (() => performance.now());
  }

  async request(
    pathAndQuery: string,
    init: Readonly<{
      method: "GET" | "POST";
      headers: HeaderInput;
      body?: Uint8Array;
    }>,
    context: ToolExecutionContext,
  ): Promise<BoundedResponse> {
    const started = this.monotonicNow();
    const operation = init.method === "GET" ? "read" : "mutation";
    const span = startSpan(this.trace, "kaji.integration.request", {
      "integration.name": this.integration,
      "integration.operation": operation,
      "http.status_family": "none",
    });
    let outcome: "success" | "error" | "cancelled" = "error";
    let scope: ReturnType<typeof requestSignal> | undefined;
    let response: FixedOriginTestResponse | undefined;
    try {
      if (this.closed) throw policyError();
      if (!this.allowedMethods.has(init.method)) throw policyError();
      if (init.body !== undefined && !(init.body instanceof Uint8Array)) throw policyError();
      const url = validatedPath(this.origin, pathAndQuery);
      const headers = validatedRequestHeaders(init.headers);
      scope = requestSignal(context, this.timeoutMs);
      response = await withAbort(
        this.transport.request(url, {
          method: init.method,
          headers,
          ...(init.body === undefined ? {} : { body: init.body }),
          signal: scope.signal,
        }),
        scope.signal,
      );
      const boundedHeaders = boundedResponseHeaders(response.headers);
      if (response.status >= 300 && response.status < 400) {
        throw new IntegrationTransportError("INTEGRATION_REDIRECT_REJECTED", "redirect_rejected");
      }
      const contentLength = boundedHeaders["content-length"];
      if (contentLength !== undefined && Number(contentLength) > this.maxResponseBytes) {
        throw responseLimit();
      }
      const iterator = response.body[Symbol.asyncIterator]();
      const chunks: Uint8Array[] = [];
      let size = 0;
      while (true) {
        const next = await withAbort(iterator.next(), scope.signal);
        if (next.done) break;
        if (!(next.value instanceof Uint8Array)) throw responseLimit();
        size += next.value.byteLength;
        if (size > this.maxResponseBytes) {
          throw responseLimit();
        }
        chunks.push(next.value);
      }
      const bytes = new Uint8Array(size);
      let offset = 0;
      for (const chunk of chunks) {
        bytes.set(chunk, offset);
        offset += chunk.byteLength;
      }
      const family = Math.floor(response.status / 100);
      span.setAttribute("http.status_family", family >= 1 && family <= 5 ? `${family}xx` : "none");
      outcome = response.status >= 200 && response.status < 300 ? "success" : "error";
      return { status: response.status, headers: boundedHeaders, bytes };
    } catch (error) {
      response?.close();
      if (error instanceof DOMException && error.name === "AbortError") outcome = "cancelled";
      span.recordError(error);
      throw error;
    } finally {
      scope?.dispose();
      recordMetric(
        this.metrics,
        "kaji.integration.request_ms",
        Math.max(0, this.monotonicNow() - started),
        { integration: this.integration, operation, outcome },
      );
      span.end();
    }
  }

  close(): void {
    if (this.closed) return;
    this.transport.close?.();
    this.closed = true;
  }
}

const productionTransport = () => new NodeHttpsTransport();

export function createGitHubRequester(
  observability: FixedOriginObservability = {},
): FixedOriginRequester & { close(): void } {
  return new FixedOriginRequesterImpl(
    { origin: new URL("https://api.github.com/"), integration: "github" },
    productionTransport(),
    observability,
  );
}

export function createGmailRequester(
  observability: FixedOriginObservability = {},
): FixedOriginRequester & { close(): void } {
  return new FixedOriginRequesterImpl(
    { origin: new URL("https://gmail.googleapis.com/"), integration: "gmail" },
    productionTransport(),
    observability,
  );
}

/** @internal Relative source tests only; absent from the package subpath. */
export function fixedOriginForTest(
  origin: string,
  transport: FixedOriginTestTransport,
  policy: Omit<Partial<FixedOriginPolicy>, "origin"> = {},
  observability: FixedOriginObservability = {},
): FixedOriginRequester & { close(): void } {
  return new FixedOriginRequesterImpl(
    { origin: new URL(origin), integration: "github", ...policy },
    transport,
    observability,
  );
}
